relayer/crypto/merkle/merkle2.go

```diff
@@ -0,0 +1,263 @@
+package merkle
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"hash"
+
+	"github.com/snowfork/go-substrate-rpc-client/v4/types"
+	"golang.org/x/crypto/sha3"
+)
+
+// Content represents the data that is stored and verified by the tree. A type that
+// implements this interface can be used as an item in the tree.
+type Content interface {
+	CalculateHash() ([]byte, error)
+	Equals(other Content) (bool, error)
+}
+
+// MerkleTree is the container for the tree. It holds a pointer to the root of the tree,
+// a list of pointers to the leaf nodes, and the merkle root.
+type MerkleTree struct {
+	Root         *Node2
+	merkleRoot   []byte
+	Leafs        []*Node2
+	hashStrategy func() hash.Hash
+	sort         bool
+}
+
+// Node represents a node, root, or leaf in the tree. It stores pointers to its immediate
+// relationships, a hash, the content stored if it is a leaf, and other metadata.
+type Node2 struct {
+	Tree   *MerkleTree
+	Parent *Node2
+	Left   *Node2
+	Right  *Node2
+	leaf   bool
+	dup    bool
+	Hash   []byte
+	C      Content
+	sort   bool
+	single bool
+}
+
+// Keccak256Content implements the Content interface provided by merkletree and represents the content stored in the tree.
+type Keccak256Content struct {
+	X types.H256
+}
+
+// CalculateHash hashes the values of a Keccak256Content
+func (t Keccak256Content) CalculateHash() ([]byte, error) {
+	return t.X[:], nil
+}
+
+// Equals tests for equality of two Contents
+func (t Keccak256Content) Equals(other Content) (bool, error) {
+	return t.X == other.(Keccak256Content).X, nil
+}
+
+// CalculateNodeHash is a helper function that calculates the hash of the node.
+func (n *Node2) CalculateNodeHash(sort bool) ([]byte, error) {
+	if n.leaf {
+		return n.C.CalculateHash()
+	}
+
+	h := n.Tree.hashStrategy()
+	if _, err := h.Write(combineTwoHash(n.Left.Hash, n.Right.Hash)); err != nil {
+		return nil, err
+	}
+
+	return h.Sum(nil), nil
+}
+
+// NewTree creates a new Merkle Tree using the content cs.
+func NewTree2(cs []Content) (*MerkleTree, error) {
+	return NewTreeWithHashStrategy(cs, sha3.NewLegacyKeccak256)
+}
+
+// NewTreeWithHashStrategy creates a new Merkle Tree using the content cs using the provided hash
+// strategy. Note that the hash type used in the type that implements the Content interface must
+// match the hash type provided to the tree.
+func NewTreeWithHashStrategy(cs []Content, hashStrategy func() hash.Hash) (*MerkleTree, error) {
+	t := &MerkleTree{
+		hashStrategy: hashStrategy,
+	}
+	root, leafs, err := buildWithContent(cs, t)
+	if err != nil {
+		return nil, err
+	}
+	t.Root = root
+	t.Leafs = leafs
+	t.merkleRoot = root.Hash
+	return t, nil
+}
+
+// GetMerklePath: Get Merkle path and indexes(left leaf or right leaf)
+func (m *MerkleTree) GetMerklePath(content Content) ([][]byte, []int64, error) {
```

Comment:
In Golang, the style isn't to use the prefix "get" for getters.

```suggestion
func (m *MerkleTree) MerklePath(content Content) ([][]byte, []int64, error) {
```